### PR TITLE
Introduce ABIv2 support into Echidna

### PR DESCRIPTION
**Pull request by incertia**
A lot of the code as it currently stands is pulled directly from HEVM with enough modification to hack it to work with ABIv2 tuples.

What works:
* Tuples classified as "static" by the ABI
* Tuples classified as "dynamic" by the ABI (probably)
* Shrinking of tuples
* Decoding `solc` output to the correct tuple type for use in the ABI
* Call sequence generation, or the hard part of what Echidna does
* Storage of return values as constants for callseq generation

What doesn't work:
* UI is broken for super long lines introduced by massive tuples

What should change:
* JSON parsing is super wonky
* Cleanup of `ABIv2.hs`

tl;dr it basically works.

### CODE REVIEW

File: lib/Echidna/ABI.hs [hunk 1]
```diff
@@ -224,15 +229,19 @@ shrinkInt x | x == -1   = pure 0
...
 shrinkAbiValue (AbiBool _)    = pure $ AbiBool False
 shrinkAbiValue (AbiBytes n b)      = AbiBytes n <$> addNulls b
 shrinkAbiValue (AbiBytesDynamic b) = fmap AbiBytesDynamic $ addNulls =<< shrinkBS b
 shrinkAbiValue (AbiString b)       = fmap AbiString       $ addNulls =<< shrinkBS b
-shrinkAbiValue (AbiArray n t l)      = AbiArray n t <$> traverse shrinkAbiValue l
+shrinkAbiValue (AbiArray n t l)    = AbiArray n t <$> traverse shrinkAbiValue l
 shrinkAbiValue (AbiArrayDynamic t l) = fmap (AbiArrayDynamic t) $ traverse shrinkAbiValue =<< shrinkV l
+shrinkAbiValue (AbiTuple v)   = AbiTuple <$> traverse shrinkAbiValue' v
+  where shrinkAbiValue' x = do
+          f <- uniform [return, shrinkAbiValue]
```
Comment (japesinator) [inline lib/Echidna/ABI.hs]: `liftM3 bool (pure x) (shrinkAbiValue x) getRandom`

Comment (japesinator) [reply lib/Echidna/ABI.hs]: let's also use this for array shrinking! it's probably useful there too

File: lib/Echidna/ABI.hs [hunk 2]
```diff
@@ -277,8 +287,9 @@ genAbiValueM = genWithDict (view constants) $ \case
   AbiStringType       -> liftM2 (\n -> AbiString       . BS.pack . take n)
                                 (getRandomR (1, 32)) getRandoms
   (AbiArrayDynamicType t) -> fmap (AbiArrayDynamic t) $ getRandomR (1, 32)
-                             >>= flip V.replicateM (genAbiValueM t)
+                              >>= flip V.replicateM (genAbiValueM t)
```
Comment (japesinator) [inline lib/Echidna/ABI.hs]: nit: deindent

File: lib/Echidna/Campaign.hs [hunk 3]
```diff
@@ -64,9 +64,9 @@ data CampaignConf = CampaignConf { testLimit     :: Int
 
 -- | State of a particular Echidna test. N.B.: \"Solved\" means a falsifying call sequence was found.
 data TestState = Open Int             -- ^ Maybe solvable, tracking attempts already made
-               | Large Int [Tx]       -- ^ Solved, maybe shrinable, tracking shrinks tried + best solve
+               | Large Int [Tx]      -- ^ Solved, maybe shrinable, tracking shrinks tried + best solve
```
Comment (japesinator) [inline lib/Echidna/Campaign.hs]: nit: fix indent

File: lib/Echidna/Campaign.hs [hunk 4]
```diff
@@ -88,7 +88,7 @@ instance ToJSON TestState where
 -- | The state of a fuzzing campaign.
 data Campaign = Campaign { _tests    :: [(SolTest, TestState)] -- ^ Tests being evaluated
                          , _coverage :: Map W256 (Set Int)     -- ^ Coverage captured (NOTE: we don't always record this)
-                         , _genDict  :: GenDict                -- ^ Generation dictionary
+                         , _genDict  :: GenDict               -- ^ Generation dictionary
```
Comment (japesinator) [inline lib/Echidna/Campaign.hs]: indent

File: lib/Echidna/Solidity.hs [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
 module Echidna.Solidity where
 
-import Control.Lens
+import Control.Lens hiding        (Indexed)
```
Comment (japesinator) [inline lib/Echidna/Solidity.hs]: why hide this?

File: lib/Echidna/Solidity.hs [hunk 6]
```diff
@@ -29,15 +29,16 @@ import Data.Text.Read             (decimal)
 import System.Process             (StdStream(..), readCreateProcess, proc, std_err)
 import System.IO                  (openFile, IOMode(..))
 
-import Echidna.ABI         (SolSignature)
+import Echidna.ABI         (SolSignature, SolSignature)
+import Echidna.ABIv2       (AbiType(..), AbiValue(..), readSolc, SolcContract,
```
Comment (japesinator) [inline lib/Echidna/Solidity.hs]: we can probably just import this entirely

File: lib/Echidna/Solidity.hs [hunk 7]
```diff
@@ -161,12 +162,11 @@ loadSpecified name cs = let ensure l e = if l == mempty then throwM e else pure
   -- Make sure everything is ready to use, then ship it
   mapM_ (uncurry ensure) $ [(abi, NoFuncs), (funs, OnlyTests)]
                         ++ if ch then [] else [(tests, NoTests)] -- ABI checks
-  ensure bc (NoBytecode $ c ^. contractName)                     -- Bytecode check
+  ensure bc (NoBytecode $ c ^. contractName)                    -- Bytecode check
```
Comment (japesinator) [inline lib/Echidna/Solidity.hs]: indent

File: lib/Echidna/Transaction.hs [hunk 8]
```diff
@@ -63,38 +63,38 @@ makeLenses 'TxConf
...
-
 -- | A contract is just an address with an ABI (for our purposes).
 type ContractA = (Addr, [SolSignature])
 
 -- | The world is made our of humans with an address, and contracts with an address + ABI.
 data World = World { _senders   :: [Addr]
                    , _receivers :: [ContractA]
                    }
-
+  deriving (Show)
```
Comment (japesinator) [inline lib/Echidna/Transaction.hs]: are we using this instance? if so, we should probably make it somewhat more clear by writing it explicitly

File: lib/Echidna/Transaction.hs [hunk 9]
```diff
@@ -63,38 +63,38 @@ makeLenses 'TxConf
...
 makeLenses ''World
 
 -- | Given generators for an origin, destination, value, and function call, generate a call
 -- transaction. Note: This doesn't generate @CREATE@s because I don't know how to do that at random.
 genTxWith :: (MonadRandom m, MonadState x m, Has World x, MonadThrow m) 
           => ([Addr] -> m Addr)                       -- ^ Sender generator
-          -> ([ContractA] -> m ContractA)             -- ^ Receiver generator
+          -> ([ContractA] -> m ContractA)           -- ^ Receiver generator
```
Comment (japesinator) [inline lib/Echidna/Transaction.hs]: indent

File: lib/Echidna/Solidity.hs (hunk 5 above)
Comment (incertia) [reply lib/Echidna/Solidity.hs]: there's an `Indexed` from HEVM that needs using somewhere.

Comment (incertia) [reply lib/Echidna/Solidity.hs]: but that was from when part of the code lived in Solidity.hs, i can get rid of this now

File: lib/Echidna/Transaction.hs (hunk 8 above)
Comment (incertia) [reply lib/Echidna/Transaction.hs]: we don't use this instance but it probably doesn't hurt to have right?

Comment (japesinator) [reply lib/Echidna/Transaction.hs]: if we're going to ship an instance, it should be a good instance. I don't think the auto-derived one is good

File: lib/Echidna/ABI.hs [hunk 10]
```diff
@@ -249,6 +256,7 @@ mutateAbiValue (AbiBytesDynamic b)   = AbiBytesDynamic   <$> mutateBS b
 mutateAbiValue (AbiString b)         = AbiString         <$> mutateBS b
 mutateAbiValue (AbiArray n t l)      = AbiArray n t      <$> traverse mutateAbiValue l
 mutateAbiValue (AbiArrayDynamic t l) = AbiArrayDynamic t <$> mutateV t l
+mutateAbiValue (AbiTuple v)  = AbiTuple <$> traverse mutateAbiValue v
```
Comment (japesinator) [inline lib/Echidna/ABI.hs]: nit: indent

File: lib/Echidna/ABI.hs [hunk 11]
```diff
@@ -279,6 +287,7 @@ genAbiValueM = genWithDict (view constants) $ \case
   (AbiArrayDynamicType t) -> fmap (AbiArrayDynamic t) $ getRandomR (1, 32)
                              >>= flip V.replicateM (genAbiValueM t)
   (AbiArrayType n t)      -> AbiArray n t <$> V.replicateM n (genAbiValueM t)
+  (AbiTupleType v)    -> AbiTuple <$> traverse genAbiValueM v
```
Comment (japesinator) [inline lib/Echidna/ABI.hs]: indent

File: lib/Echidna/ABI.hs [hunk 12]
```diff
@@ -1,4 +1,5 @@
 {-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
```
Comment (japesinator) [inline lib/Echidna/ABI.hs]: is this used?

Comment (incertia) [reply lib/Echidna/ABI.hs]: this can get deleted

**Pull request merged**